exporters/metrics/src/main/java/com.google.cloud.opentelemetry.metric/MetricTranslator.java

```diff
@@ -0,0 +1,123 @@
+  }
+
+  static MetricDescriptor mapMetricDescriptor(MetricData metric) {
+    String instrumentName = metric.getDescriptor().getName();
+    MetricDescriptor.Builder builder = MetricDescriptor.newBuilder().setDisplayName(instrumentName)
+        .setType(mapMetricType(instrumentName));
+    metric.getDescriptor().getConstantLabels().forEach((key, value) -> builder.addLabels(mapConstantLabel(key, value)));
+
+    MetricData.Descriptor.Type metricType = metric.getDescriptor().getType();
+    if (GAUGE_TYPES.contains(metricType)) {
+      builder.setMetricKind(MetricDescriptor.MetricKind.GAUGE);
+    } else if (CUMULATIVE_TYPES.contains(metricType)) {
+      builder.setMetricKind(MetricDescriptor.MetricKind.CUMULATIVE);
+    } else {
+      logger.error("Metric type {} not supported. Only gauge and cumulative types are supported.", metricType);
+      return null;
+    }
+    if (LONG_TYPES.contains(metricType)) {
+      builder.setValueType(MetricDescriptor.ValueType.INT64);
+    } else if (DOUBLE_TYPES.contains(metricType)) {
+      builder.setValueType(MetricDescriptor.ValueType.DOUBLE);
+    } else {
+      logger.error("Metric type {} not supported. Only long and double types are supported.", metricType);
+      return null;
+    }
+    return builder.build();
+  }
+
+  private static String mapMetricType(String instrumentName) {
+    for (String domain : KNOWN_DOMAINS) {
+      if (instrumentName.contains(domain)) {
+        return instrumentName;
+      }
+    }
+    return DESCRIPTOR_TYPE_URL + instrumentName;
+  }
+
+  static LabelDescriptor mapConstantLabel(String key, String value) {
+    LabelDescriptor.Builder builder = LabelDescriptor.newBuilder().setKey(key);
+    if ("true".equalsIgnoreCase(value) || "false".equalsIgnoreCase(value)) {
+      builder.setValueType(LabelDescriptor.ValueType.BOOL);
+    } else if (Longs.tryParse(value) != null) {
+      builder.setValueType(LabelDescriptor.ValueType.INT64);
+    } else {
+      builder.setValueType(LabelDescriptor.ValueType.STRING);
+    }
+    return builder.build();
+  }
+
+  static Point mapPoint(MetricData metric, MetricData.Point point, MetricWithLabels updateKey,
+      Map<MetricWithLabels, Long> lastUpdatedTime) {
+    Builder pointBuilder = Point.newBuilder();
+    Type type = metric.getDescriptor().getType();
+    if (LONG_TYPES.contains(type)) {
+      pointBuilder.setValue(TypedValue.newBuilder().setInt64Value(((MetricData.LongPoint) point).getValue()));
+    } else if (DOUBLE_TYPES.contains(type)) {
+      pointBuilder.setValue(TypedValue.newBuilder().setDoubleValue(((MetricData.DoublePoint) point).getValue()));
+    } else {
+      logger.error("Type {} not supported", type);
+      return null;
+    }
+    pointBuilder.setInterval(mapInterval(point));
+    lastUpdatedTime.put(updateKey, point.getEpochNanos());
+    return pointBuilder.build();
+  }
+
+  static TimeInterval mapInterval(MetricData.Point point) {
+    Timestamp startTime = Timestamp.newBuilder()
```

Comment:
I think you may still need to add up to 1ms delta for the "fake deltas"/Counter instruments, to guarantee that the reset intervals are at least 1ms apart. For the SumObserver instruments (real cumulatives), the current code here should be fine since start will just be the same as the previous export pass.

Or maybe a more general approach would be to always subtract 1ms from the end times, which I think would work for both cases? Idk, anyone else have thoughts on this?

Reply:
Currently those types (NON_MONOTONIC_LONG, MONOTONIC_DOUBLE, etc) are the only information we get in Java, and it's not possible to tell apart the specific instruments.
I removed the delta part because it seems like https://github.com/open-telemetry/opentelemetry-java/blob/d42d4cd9e2d79c8e5bdc3c43aac5e077f4147794/exporters/otlp/src/main/java/io/opentelemetry/exporters/otlp/MetricAdapter.java#L155 is treating all of them (apart from SUMMARY which I'm ignoring) as cumulative.
WDYT?